src/finchlite/finch_assembly/dataflow.py

```diff
@@ -0,0 +1,150 @@
+            case _:
+                return str(expr)
+
+    @abstractmethod
+    def print_lattice_value(self, name, state: dict) -> str:
+        """Format the lattice value associated with a variable for annotation."""
+        ...
+
+
+class AssemblyCopyPropagation(AbstractAssemblyDataflow):
+    def direction(self) -> str:
+        """
+        Copy propagation is a forward analysis.
+        """
+        return "forward"
+
+    def transfer(self, stmts, state: dict) -> dict:
+        new_state = state.copy()
+
+        for stmt in stmts:
+            match stmt:
+                case Assign(lhs, rhs):
+                    # Get the variable name being assigned to
+                    var_name = self._get_variable_name(lhs)
+
+                    if var_name is not None:
+                        new_state[var_name] = rhs
+
+                        # invalidate any copies that point
+                        # to the variable being assigned
+                        to_remove = []
+                        for var, val in new_state.items():
+                            if val == lhs:
+                                to_remove.append(var)
+
+                        for var in to_remove:
+                            new_state.pop(var)
+
+        return new_state
+
+    def join(self, state_1: dict, state_2: dict) -> dict:
+        result = {}
+
+        # only keep copy relationships that exist in both states with the same value
+        for var_name in state_1:
+            if var_name in state_2 and self._values_equal(
+                state_1[var_name], state_2[var_name]
+            ):
+                result[var_name] = state_1[var_name]
+
+        return result
+
+    def print_lattice_value(self, name, state: dict) -> str:
+        visited: set[str] = set()
+        value = state.get(name)
+
+        # iterate through states until we reach a Literal or a variable
+        # without a mapping (root value)
+        while True:
+            match value:
+                case Literal(v):
+                    return qual_str(v)
+                case TaggedVariable(Variable(next_name, _), _) | Variable(next_name, _):
+                    # lattice was was already visited -> return
+                    if next_name in visited:
+                        return str(value)
+
+                    visited.add(next_name)
+                    next_val = state.get(next_name)
+
+                    # reached root value -> return
+                    if next_val is None:
+                        return str(value)
+
+                    value = next_val
+                case _:
+                    return str(value)
+
+    def _get_variable_name(self, var) -> str | None:
+        match var:
+            case Variable(name, _):
+                return name
+            case TaggedVariable(Variable(name, _), _):
+                return name
+            case _:
+                return None
+
+    def _values_equal(self, val1, val2) -> bool:
```

Comment:
I'd rather the algorithm be written carefully to know whether the values in dictionaries are Literals, Variables, or taggedVariables. If all that matters is the var, I'd like to see tagged_var.var == tagged_var2.var